packages/bitcoin-wallet-provider/lib/BitcoinWalletProvider.ts

```diff
@@ -57,6 +63,168 @@ export default class BitcoinWalletProvider
+    const p2wsh = bitcoin.payments.p2wsh({
+      redeem: p2ms,
+      network: this._network,
+    });
+
+    return p2wsh;
+  }
+
+  /**
+   * Creates a native-segwit multi-signature address (P2MS in P2WSH) with n signatures of m required keys
+   * https://developer.bitcoin.org/reference/rpc/createmultisig.html
+   * @param m the number of required signatures
+   * @param pubkeys n possible pubkeys in total
+   * @returns a json object containing the `address` and `redeemScript`
+   */
+  createMultisig(m: number, pubkeys: string[]): CreateMultisigResponse {
+    const p2wsh = this._createMultisigPayment(m, pubkeys);
+    return {
+      address: p2wsh.address,
+      redeemScript: p2wsh.redeem?.output?.toString('hex'),
+    };
+  }
+
+  /**
+   * Creates a PSBT of a native-segwit multi-signature address (P2MS in P2WSH) with n signatures of m required keys
+   * https://developer.bitcoin.org/reference/rpc/createmultisig.html
+   * https://developer.bitcoin.org/reference/rpc/createpsbt.html
+   * @param m the number of required signatures
+   * @param pubkeys n possible pubkeys in total
+   * @param inputs the Inputs to the PSBT
+   * @param ouputs the Outputs to the PSBT
+   * @returns a base64 encoded psbt string
+   */
+  buildMultisigPSBT(
+    m: number,
+    pubkeys: string[],
+    inputs: Input[],
+    outputs: Output[],
+  ): string {
+    assert(inputs.length > 0, 'no inputs found');
+    assert(outputs.length > 0, 'no outputs found');
+
+    const p2wsh = this._createMultisigPayment(m, pubkeys);
+
+    // Verify pubkeyhash for all inputs matches the p2wsh hash
+    assert(
+      inputs.every(
+        (input: Input) => p2wsh.output.toString('hex') === input.scriptPubKey,
+      ),
+      'address pubkeyhash does not match input scriptPubKey',
+    );
+
+    // creator
+    const psbt = new bitcoin.Psbt({ network: this._network });
+
+    // updater
+    inputs.forEach((input: Input) => {
+      psbt.addInput({
+        hash: input.txid,
+        index: input.vout,
+        witnessUtxo: { script: p2wsh.output, value: input.value },
+        witnessScript: p2wsh.redeem.output,
+      });
+    });
+
+    outputs.forEach((output: Output) => {
+      psbt.addOutput({
+        address: output.to,
+        value: output.value,
+      });
+    });
+
+    return psbt.toBase64();
+  }
+
+  /**
+   * Update a PSBT with input information from our wallet and then sign inputs that we can sign for
+   * https://developer.bitcoin.org/reference/rpc/walletprocesspsbt.html
+   * @param psbt a base64 encoded psbt string
+   * @returns a base64 encoded signed psbt string
+   */
+  async walletProcessPSBT(psbtString: string): Promise<string> {
+    const psbt = bitcoin.Psbt.fromBase64(psbtString);
+
+    await Promise.all(
+      psbt.data.inputs.map(async (input, i: number) => {
+        const scriptStack = bitcoin.script.decompile(input.witnessScript);
+        const pubkeys = scriptStack.filter((data) => Buffer.isBuffer(data));
```

Comment:
There could be non-pubkey data that's a Buffer, could do with an additional check with `secp256k1.publicKeyVerify(pubkey)`

Reply:
Added!